File: esto-angular-frontend/check_html.py
import re
import sys

def check_html():
    try:
        with open('c:\\esto-platform\\esto-angular-frontend\\src\\app\\pages\\admin\\messages\\admin-messages.component.ts', 'r', encoding='utf8') as f:
            content = f.read()
    except Exception as e:
        print("Error reading file:", e)
        return

    match = re.search(r'template:\s*`([\s\S]*?)`', content)
    if not match:
        print("No template found")
        return
    
    html = match.group(1)
    lines = html.split('\n')
    
    stack = []
    self_closing = ['img', 'input', 'br', 'hr', 'path', 'circle']
    
    for i, line in enumerate(lines):
        # find all tags <tag> or </tag>
        # regex to match <name attributes> or </name>
        tags = re.finditer(r'<\/?([a-zA-Z0-9-]+)[^>]*>', line)
        for tag_match in tags:
            full_tag = tag_match.group(0)
            tag_name = tag_match.group(1).lower()
            
            if tag_name in self_closing:
                continue
            
            if full_tag.endswith('/>'):
                continue
                
            if full_tag.startswith('</'):
                if not stack:
                    print(f"Error at line {i+1}: Unexpected </{tag_name}>, stack is empty. Line: {line}")
                    continue
                last = stack.pop()
                if last['name'] != tag_name:
                    print(f"Error at line {i+1}: Expected </{last['name']}> (from line {last['line']}), but found </{tag_name}>. Line: {line}")
            else:
                stack.append({'name': tag_name, 'line': i+1})
                
    if stack:
        print("Unclosed tags remaining:")
        for t in stack:
            print(f"<{t['name']}> at line {t['line']}")
    else:
        print("All tags balanced!")

check_html()

File: esto-angular-frontend/check_html.js
const fs = require('fs');
const content = fs.readFileSync('c:\\esto-platform\\esto-angular-frontend\\src\\app\\pages\\admin\\messages\\admin-messages.component.ts', 'utf8');
const templateMatch = content.match(/template:\s*`([\s\S]*?)`/);
if (!templateMatch) {
  console.log("No template found");
  process.exit(1);
}
const html = templateMatch[1];
const stack = [];
const lines = html.split('\n');
const selfClosing = ['img', 'input', 'br', 'hr', 'path', 'circle'];

for (let i = 0; i < lines.length; i++) {
  const line = lines[i];
  const regex = /<\/?([a-zA-Z0-9-]+)[^>]*>/g;
  let match;
  while ((match = regex.exec(line)) !== null) {
    const fullTag = match[0];
    const tagName = match[1].toLowerCase();
    
    if (selfClosing.includes(tagName)) continue;
    
    // Check if it ends with "/>"
    if (fullTag.endsWith('/>')) continue;
    
    if (fullTag.startsWith('</')) {
      if (stack.length === 0) {
        console.log(`Error at line ${i + 1}: Unexpected </${tagName}>, stack is empty`);
        continue;
      }
      const last = stack.pop();
      if (last.tagName !== tagName) {
        console.log(`Error at line ${i + 1}: Expected </${last.tagName}> to close tag from line ${last.line}, but found </${tagName}>`);
        console.log(line);
      }
    } else {
      stack.push({tagName, line: i + 1});
    }
  }
}
if (stack.length > 0) {
  console.log("Unclosed tags remaining:");
  console.log(stack);
} else {
  console.log("All tags balanced!");
}
